### ISSUE TITLE
Redesign reader() and writer()

### ISSUE THREAD

**Issue by stjepang** [stjepang/smol#79]
Functions `reader()` and `writer()` return `impl Trait` types, which is not great if we want to store a value of that type somewhere. Perhaps it'd be better to return `Reader` and `Writer` types that we define in this crate.

Furthermore, it would be nice to have some sugar around common patterns. Writing into a file currently looks like this:

```rust
let file = blocking!(File::open("foo.txt"))?;
let mut file = writer(file);
```

But what if we could do this instead?

```rust
let mut file = Writer::file("foo.txt").await?;
```

If we extend this idea a little bit, what if we had the following API?

```rust
struct Reader;
struct Writer;

impl Reader {
    fn blocking(r: impl Read + Send + 'static) -> Reader;
    fn stdin() -> Reader;
    async fn file(p: impl AsRef<Path>) -> io::Result<Reader>
    async fn open(opts: OpenOptions) -> io::Result<Reader>;
}

impl Writer {
    fn blocking(w: impl Write + Send + 'static) -> Writer;
    fn stdout() -> Writer;
    fn stderr() -> Writer;
    async fn file(path: impl AsRef<Path>) -> io::Result<Writer>;
    async fn open(opts: OpenOptions) -> io::Result<Writer>;
}
```

If we rename `Async::new()` to `Async::pollable()`, then it'd also make sense to use `Reader::blocking()` rather than `Reader::new()` to indicate what types the reader (or writer) should be used with.

Previous discussion: https://github.com/stjepang/smol/issues/36

**Comment by stjepang** [comment]
Almost forgot to mention, the same applies to `iter()`:

```rust
struct Iter<T>;

impl Iter<T> {
    fn blocking(iter: impl IntoIterator<Item = T>) -> Iter<T>;
}

impl Iter<ReadDir> {
    async fn read_dir(path: impl AsRef<Path>) -> io::Result<Iter<ReadDir>>;
}
```

Although I'm torn between names `Iter` and `Streamer`...

**Comment by jjl** [comment]
Given that I previously mentioned the awkwardness of being unable to write the type of `Sender::send`, you might find it interesting to note,  that I've just given up and started using the `type_alias_impl_trait` feature lately.

That said, this does look good.

**Comment by stjepang** [comment]
@jjl I have an idea for a type similar to `Async<T>` that uses the blocking executor to do I/O.

Imagine you could this:

```rust
let mut s = String::new();
let file = Async::<File>::open("foo.txt").await?;
file.read_to_end(&mut s).await?;

Async::copy("a.txt", "b.txt").await?;
```

Except I don't know what to name this new type - it must be something different from `Async` :)
`Unblock<T>`? `Blocking<T>`? `Threaded<T>`? `Pooled<T>`? `Unsync<T>`?

**Comment by zicklag** [comment]
> Writing into a file currently looks like this:

```rust
let file = blocking!(File::open("foo.txt"))?;
let mut file = writer(file);
```

I actually don't think that that is bad at all. The thing that I'm not sure about immediately with having a function like `File::open` for the `Writer` is what about the situations where you want to open a file with different options? I usually use `OpenOptions` for opening files.

That said, maybe it couldn't hurt and it would help other people get started. I'm not sure where I stand on it yet, but there's a thought. :)

**Comment by jjl** [comment]
@stjepang My immediate thought before reading your suggestions list was 'Blocking', so I was pleased to see it in the list :)

I'd expect that you'd want to be able to spawn it against a thread pool though. I'm not sure how this would play with that.

**Comment by stjepang** [comment]
For now, I'm calling this type `Bio<T>`, where BIO stands for Blocking I/O. Think of it as blocking I/O done asynchronously.

Here's a possible API:

```rust
impl Bio<T> {
    fn new(t: T) -> Bio<T>;
}

impl Bio<Stdin> {
    fn stdin() -> Bio<Stdin>;
}

impl Bio<Stdout> {
    fn stdout() -> Bio<Stdout>;
}

impl Bio<Stderr> {
    fn stderr() -> Bio<Stderr>;
}

impl Bio<File> {
    async fn open(path: AsRef<Path>) -> io::Result<Bio<File>>;
    async fn create(path: AsRef<Path>) -> io::Result<Bio<File>>;
    async fn with_options(opts: OpenOptions) -> io::Result<Bio<File>>;
}

impl<T: Read> AsyncRead for Bio<T> {}
impl<T: Write> AsyncWrite for Bio<T> {}

impl Future for Bio<T> {
    type Output = T;
}
```

**Comment by jjl** [comment]
I don't like the name, but the API looks sensible enough. It does make me wonder what to do about cpu-bound blocking tasks, however.

**Comment by stjepang** [comment]
@jjl Okay let's go with `Blocking<T>` then :)

We could also add some functions like these:

```rust
impl Blocking<()> {
    async fn read_to_string(path: AsRef<Path>) -> io::Result<String>;
    async fn metadata(path: AsRef<Path>) -> io::Result<Metadata>;
}
```

> It does make me wonder what to do about cpu-bound blocking tasks, however.

It would be possible to add something like:

```rust
impl<T> Blocking<T> {
    fn compute(f: impl FnOnce() -> T + Send + 'static) -> Blocking<T>;
}
```

**Comment by jjl** [comment]
I'm now in two minds :)

On the one hand, I/O is different enough from CPU that we have the traits for it, on the other, from the executor's perspective, they're just threads that do stuff.

So I'm wondering whether it makes sense to push it into one type because of the latter, or not, because of the former. Although perhaps a careful choice of blanket implementation would suffice? I don't know, I'm tired :)

That sad,

**Comment by Ekleog** [comment]
I kind of think it'd make sense to somehow separate between IO-bound and CPU-bound blocking operations: while it makes sense to have 500 IO-bound tasks at the same time, if one has 500 CPU-bound tasks at the same time, I'd guess performance is going to take a hit?

**Comment by jjl** [comment]
It depends on how you're looking at it. Some people spawn cpu-bound and io-bound tasks separately and others bundle them together, having just low-latency/high-latency pools.

I think this could be solved by providing a pool for it to execute against when creating the `Blocking`.

**Comment by stjepang** [comment]
We'll replace `iter()`, `reader()`, `writer()` and `Task::blocking()` with `Blocking<T>`:

https://docs.rs/blocking/0.3.1/blocking/

**Comment by zicklag** [comment]
That's awesome! And because it spawns it's own thread pool it should be compatible with any async executor right?

Also that means that we don't need those components in smol, which might make smol a little smaller for people who don't need blocking support? Either way, that is sweet!

**Comment by stjepang** [comment]
Exactly - `blocking` is compatible with all executors. :)

I'm thinking of re-exporting `Blocking` type from smol, however. It's a rather small dependency and doesn't pull in any transitive dependencies which smol doesn't already have.

**Comment by ririsoft** [comment]
Hello,

I am missing `std::io::Seek` trait when using `std::fs::File` with `smol`. Do you plan to address this with this redesign ? Otherwise is `async-std` the long term way to go for such feature ?

Thanks in advance for your comments here.
Cheers (and congratulation for the awesome work on async).

**Comment by stjepang** [comment]
@ririsoft Over time, `smol` with replace its blocking executor with the `blocking` crate. I suggest using `Unblock<File>`: https://docs.rs/blocking/0.4.4/blocking/struct.Unblock.html

To seek, you could do `file.with_mut(|f| f.seek(pos)).await?`. Perhaps it'd also be useful to implement `AsyncSeek` on `Unblock` so that you could do `file.seek(pos).await?`...

**Comment by ririsoft** [comment]
> Over time, `smol` with replace its blocking executor with the `blocking` crate. I suggest using `Unblock<File>`: https://docs.rs/blocking/0.4.4/blocking/struct.Unblock.html

I will thanks !

> To seek, you could do `file.with_mut(|f| f.seek(pos)).await?`. Perhaps it'd also be useful to implement `AsyncSeek` on `Unblock` so that you could do `file.seek(pos).await?`...

Indeed it would be really nice if  `Unblock<T>` implements `AsyncSeek` if `T` implements `Seek`.
Should I open a tracking issue for that on the `blocking` project or it is already on your todo list ?

**Comment by stjepang** [comment]
You can open an issue there :)

**Comment by stjepang** [comment]
Closing this - we should now just deprecate `reader()` and `writer()` in favor of `blocking`.
